target-libretro/libretro.cpp

```diff
@@ -682,19 +746,95 @@ void retro_get_system_info(struct retro_system_info *info) {
   info->need_fullpath    = false;
 }
 
-void retro_get_system_av_info(struct retro_system_av_info *info) {
+#define SAMPLE_FREQ_PAL 14750000.0
+#define SAMPLE_FREQ_NTSC (135000000.0/11.0)
+static double get_aspect_ratio(unsigned width, unsigned height)
+{
+  bool is_pal = retro_get_region() == RETRO_REGION_PAL;
+  if (core_bind.aspect_ratio_mode == 1)
+    is_pal = false; 
+  else if (core_bind.aspect_ratio_mode == 2)
+    is_pal = true;
+
+  double par = (is_pal ? SAMPLE_FREQ_PAL : SAMPLE_FREQ_NTSC) / (SuperFamicom::ppu.frequency / 2.0);
+  return (double)width * par / (double)height; 
+}
+
+static void get_system_av_info(struct retro_system_av_info *info)
+{
+  update_variables();
+  memset(info, 0, sizeof(retro_system_av_info));
   struct retro_system_timing timing = { 0.0, 32040.5 };
   timing.fps = retro_get_region() == RETRO_REGION_NTSC ? 21477272.0 / 357366.0 : 21281370.0 / 425568.0;
+  unsigned max_width = 512;
+  unsigned max_height = core_bind.crop_overscan ? 448 : 478;
+  unsigned base_width = 256;
+  unsigned base_height = SuperFamicom::ppu.overscan() ? 239 : 224;
 
-  if (!core_bind.penviron(RETRO_ENVIRONMENT_GET_OVERSCAN, &core_bind.overscan))
-     core_bind.overscan = false;
+  if (core_bind.crop_overscan)
+    base_height = 224;
 
-  unsigned base_width = 256;
-  unsigned base_height = core_bind.overscan ? 240 : 224;
-  struct retro_game_geometry geom = { base_width, base_height, base_width << 1, base_height << 1, 4.0 / 3.0 };
+  double aspect_ratio = get_aspect_ratio(base_width, base_height);
+#ifdef __DEBUG
+  output(RETRO_LOG_INFO, "Base height: %u\n", base_height);
+  output(RETRO_LOG_INFO, "Base width: %u\n", base_width);
+  output(RETRO_LOG_INFO, "Aspect ratio: %f\n", aspect_ratio);
+  output(RETRO_LOG_INFO, "FPS: %f\n", timing.fps);
+#endif
+  struct retro_game_geometry geom = { base_width, base_height, max_width, max_height, (float)aspect_ratio };
 
-  info->timing   = timing;
+  info->timing = timing;
   info->geometry = geom;
+}
+
+retro_system_av_info current_info;
+
+static void update_system_av_info()
+{
+  struct retro_system_av_info info;
+  get_system_av_info(&info);
+  bool b = false;
+  if (current_info.geometry.aspect_ratio != info.geometry.aspect_ratio)
+    b = true;
```

Comment:
b |= (current_info.geometry.aspect_ratio != info.geometry.aspect_ratio);

or maybe just dump them all inside that if.